File: crates/yokan/demo/trend.py
# /// script
# requires-python = ">=3.14"
# ///
"""Chart data from a list[float] cell — the float rule in action:

float TEXT is out of the dialect (str(2.0) diverges across tiers),
but float DATA is fine: both tiers dump chart values through the
same kernel renderer, and the gate proves it.
"""
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

import yokan as ui  # noqa: E402
from yokan import State  # noqa: E402

values: State[list[float]] = State([3.0, 5.0, 2.0])
limit: State[float] = State(4.5)


def bump():
    values.set(values() + [8.0])


def raise_limit():
    limit.set(limit() + 0.5)


def view():
    with ui.column(spacing=10, padding=14):
        ui.text(f"points: {len(values())}", size=14)
        ui.line_chart(values(), height=120.0)
        ui.bar_chart(values(), height=90.0)
        ui.text(f"limit: {limit():.1f}", size=12, color="#8a8f98")
        with ui.row(spacing=8):
            ui.button("add point", on_click=bump)
            ui.button("raise limit", on_click=raise_limit)


if __name__ == "__main__":
    ui.run(view, title="trend")

File: gomamochi/demo/trend.go
// One list of numbers, drawn twice. A chart takes its data as its first
// argument, so a view that computes the numbers reads in order.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

type Trend struct {
	values []float64
	limit  float64
}

func (t *Trend) bump() {
	t.values = append(t.values, 8)
}

func (t *Trend) View() Element {
	return Column(
		Text(fmt.Sprintf("points: %d", len(t.values))).Size(14),
		LineChart(t.values).Height(120),
		BarChart(t.values).Height(90),
		Text(fmt.Sprintf("limit: %.1f", t.limit)).Size(12).Color("#8a8f98"),
		Row(
			Button("add point").OnClick(func() { t.bump() }),
			Button("raise limit").OnClick(func() { t.limit += 0.5 }),
		).Spacing(8),
	).Spacing(10).Padding(14)
}

func main() {
	Run(&Trend{values: []float64{3, 5, 2}, limit: 4.5}, Title("trend"))
}
